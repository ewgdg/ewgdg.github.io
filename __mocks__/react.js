const React = jest.requireActual("react")
module.exports = {
  ...React,
  useContext: jest.fn().mockImplementation(React.useContext),
}
