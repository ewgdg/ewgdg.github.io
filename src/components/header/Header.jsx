/* eslint-disable react/destructuring-assignment */

import PropTypes from "prop-types"
import React, { useContext, useEffect, useState } from "react"
import AppBar from "@material-ui/core/AppBar"
import Button from "@material-ui/core/Button"
import Toolbar from "@material-ui/core/Toolbar"
import IconButton from "@material-ui/core/IconButton"

import Typography from "@material-ui/core/Typography"
import Box from "@material-ui/core/Box"

import Fab from "@material-ui/core/Fab"
import KeyboardArrowUpIcon from "@material-ui/icons/KeyboardArrowUp"
import Zoom from "@material-ui/core/Zoom"

import { navigate } from "gatsby"

import LayoutContext from "../../contexts/LayoutContext"
import { scrollIntoView, ScrollDetector } from "../../utilities/scroll"
import { clearHistoryState } from "../../contexts/useRestoreComponentState"
import useLayoutContext from "../../contexts/useLayoutContext"

function useScrollTrigger({ threshold, target }) {
  const [trigger, setTrigger] = useState(false)

  useEffect(() => {
    const scene = new ScrollDetector({
      scrollLayer: target,
      triggerHook: 0,
      offset: threshold,
    })

    scene.setEventListener(progress => {
      if (progress > 0) {
        setTrigger(true)
      } else {
        setTrigger(false)
      }
    })

    return () => {
      scene.destroy()
      setTrigger(false)
    }
  }, [threshold, target])
  return trigger
}
function BackToTop(props) {
  const { children, anchorId } = props
  // const classes = useStyles();
  const context = useContext(LayoutContext)
  const trigger = useScrollTrigger({
    threshold: 100,
    target: context.scrollLayer,
  })

  const handleClick = event => {
    const anchor = (event.target.ownerDocument || document).querySelector(
      `#${anchorId}`
    )
    if (anchor) {
      scrollIntoView(anchor, context.scrollLayer, 888)
      // anchor.scrollIntoView({ behavior: "smooth", block: "start" })
    }
  }

  return (
    <Zoom in={trigger}>
      <div onClick={handleClick} role="presentation" className="scroll-button">
        {children}
        <style jsx>
          {`
            .scroll-button {
              position: fixed;
              bottom: 2rem;
              right: 2rem;
            }
          `}
        </style>
      </div>
    </Zoom>
  )
}
BackToTop.propTypes = {
  children: PropTypes.element.isRequired,
  anchorId: PropTypes.string.isRequired,
}

const Header = ({ position, color, style }) => {
  const context = useLayoutContext()
  return (
    <div>
      <AppBar
        position={position}
        id="navbar-top"
        className="navbar"
        style={style}
      >
        <Toolbar style={{ justifyContent: "space-around" }} variant="dense">
          <IconButton
            edge="start"
            color="inherit"
            aria-label="menu"
            onClick={() => {
              clearHistoryState(["/"], context)
              navigate("/")
            }}
          >
            {/* <MenuIcon /> */}
            <Typography variant="h6" className="title">
              Xian
            </Typography>
          </IconButton>

          <Button
            onClick={() => {
              navigate("/about")
            }}
            color="inherit"
          >
            <Box display="flex" flexDirection="row">
              <Typography variant="h6" className="title">
                About
              </Typography>
            </Box>
          </Button>

          <Button
            onClick={() => {
              clearHistoryState(["/blog"], context)
              navigate("/blog")
            }}
            color="inherit"
          >
            Blog
          </Button>
        </Toolbar>

        <BackToTop anchorId="navbar-top">
          <Fab color="secondary" size="small" aria-label="scroll back to top">
            <KeyboardArrowUpIcon />
          </Fab>
        </BackToTop>
      </AppBar>
      <style jsx global>
        {`
          .navbar {
            background-color: transparent;
            color: ${color};
            user-select: none;
          }
        `}
      </style>
    </div>
  )
}

Header.propTypes = {
  position: PropTypes.string,
  color: PropTypes.string,
}
Header.defaultProps = {
  position: "static",
  color: "black",
}
export default Header
