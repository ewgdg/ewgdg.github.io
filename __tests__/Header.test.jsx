/* eslint-disable react/jsx-filename-extension */
import React from "react"
// import renderer from "react-test-renderer"
import { render, fireEvent } from "@testing-library/react"
// import { createRender } from "@material-ui/core/test-utils"
import Header from "../src/components/header/Header"

describe("Header", () => {
  test("renders correctly", () => {
    const tree = render(<Header position="absolute" />)
    expect(tree).toMatchSnapshot()
  })

  test("properly navigate upon click about link", () => {
    const { getByText } = render(<Header />)
    fireEvent.click(getByText(/about/i))
    expect(window.location.href).toBe("/about")
  })
})
