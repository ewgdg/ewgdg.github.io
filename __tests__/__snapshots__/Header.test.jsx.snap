// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <header
          class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionAbsolute navbar MuiAppBar-colorPrimary"
          id="navbar-top"
        >
          <div
            class="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
            style="justify-content: space-around;"
          >
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <h6
                  class="MuiTypography-root title MuiTypography-h6"
                >
                  Xian
                </h6>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <div
                  class="MuiBox-root MuiBox-root-84"
                >
                  <h6
                    class="MuiTypography-root title MuiTypography-h6"
                  >
                    About
                  </h6>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <h6
                  class="MuiTypography-root title MuiTypography-h6"
                >
                  Blog
                </h6>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="scroll-button"
            role="presentation"
            style="transform: scale(0); visibility: hidden;"
          >
            <button
              aria-label="scroll back to top"
              class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-secondary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiFab-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <style>
              
            .scroll-button {
              position: fixed;
              bottom: 2rem;
              right: 2rem;
            }
          
            </style>
          </div>
          <div
            style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          >
            <svg
              aria-haspopup="true"
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-104"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM6 9h12v2H6V9zm8 5H6v-2h8v2zm4-6H6V6h12v2z"
              />
            </svg>
          </div>
        </header>
        <style>
          
          .navbar {
            background-color: transparent;
            color: black;
            user-select: none;
          }
        
        </style>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionAbsolute navbar MuiAppBar-colorPrimary"
        id="navbar-top"
      >
        <div
          class="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
          style="justify-content: space-around;"
        >
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <h6
                class="MuiTypography-root title MuiTypography-h6"
              >
                Xian
              </h6>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <div
                class="MuiBox-root MuiBox-root-84"
              >
                <h6
                  class="MuiTypography-root title MuiTypography-h6"
                >
                  About
                </h6>
              </div>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <h6
                class="MuiTypography-root title MuiTypography-h6"
              >
                Blog
              </h6>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="scroll-button"
          role="presentation"
          style="transform: scale(0); visibility: hidden;"
        >
          <button
            aria-label="scroll back to top"
            class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-secondary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <style>
            
            .scroll-button {
              position: fixed;
              bottom: 2rem;
              right: 2rem;
            }
          
          </style>
        </div>
        <div
          style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <svg
            aria-haspopup="true"
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-icon-104"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM6 9h12v2H6V9zm8 5H6v-2h8v2zm4-6H6V6h12v2z"
            />
          </svg>
        </div>
      </header>
      <style>
        
          .navbar {
            background-color: transparent;
            color: black;
            user-select: none;
          }
        
      </style>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
